src/Uno.UWP/Helpers/DrawableHelper.Android.cs

```diff
@@ -78,13 +75,43 @@ public static Drawable FromUri(Uri uri)
 
 		private static void InitializeDrawablesLookup()
 		{
-			_drawablesLookup = _drawables
-				.GetFields(BindingFlags.Static | BindingFlags.Public)
+			_drawablesLookup = _drawables.GetFields(BindingFlags.Static | BindingFlags.Public)
 				.ToDictionary(
 					p => p.Name,
 					p => (int)p.GetValue(null)
 				);
 		}
+
+		private static int? GetResourceId(string resourceKey, Func<string, int> _drawableLookupProvider = null)
+		{
+			if (_drawablesLookup == null)
+			{
+				throw new InvalidOperationException("Drawable resources were not initialized. "
+					+ "On Android, local assets are only available after App.InitializeComponent() has been called.");
+			}
+
+			var id = _drawableLookupProvider == null
+				? _drawablesLookup.UnoGetValueOrDefault(resourceKey)
+				: _drawableLookupProvider(resourceKey);
+
+			if (id == 0)
+			{
+				if (typeof(DrawableHelper).Log().IsEnabled(Microsoft.Extensions.Logging.LogLevel.Error))
+				{
+					typeof(DrawableHelper).Log().Error($"Couldn't find drawable with key: {resourceKey}");
+				}
+
+				return null;
+			}
+
+			return id;
+		}
+
+		private static int LookupDrawableFromFileName(string fileName) =>
+			_drawablesLookup
```

Comment:
Not my favourite solution but we can't really build a second dictionary here since the filenames themselves can include underscores

Reply:
Indeed. And this may become a performance issue... I wonder if we should instead a source generator to build the dictionaries. After all, enumerating the entries using reflection is not a good thing to do. This way, we could provide both lookups in an efficient way.

The generator would create a class that would provide two lookup methods with a switch/case based on the same Resource class used here: 
https://github.com/unoplatform/uno/blob/0c9de5d418ab197928bda5a69c61670298a18652/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs#L475-L477

Reply:
Not sure I understand fully, would we be creating a dictionary using the source generators or would the source generators be creating two switch case methods that would be "hardcoded' with all the drawable names?

Reply:
This would be pretty much the same, as switch/case statements big enough get converted to dictionaries by the compiler. Are you familiar with source generation ? You can find a simple generator here: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.UITests.Generator/SnapShotTestGenerator.cs

Reply:
Source generation is pretty new to me, I'll give it some research and come back to this

Reply:
Ok, bear with me here. So what I should do is create an implementation of a `SourceGenerator`, something like `DrawableLookupGenerator`. Within the `Execute` method, I am given a `SourceGeneratorContext` with which I can grab the `Compilation` of the app. I can then query that `Compilation` to find the global `Resource.Drawable` class and query for its public properties. Now that I have the public properties, I can build up a massive switch statement method that will match up a provided file name or file path to the proper property name? I'd create two lookup methods, one that accepts just a straight filename that would match to potentially more than one property name of the `Resource.Drawable` class, we could search based on if the property names end with the filename. And one that accepts a path which we can encode and then see which `Resource.Drawable` property name matches that encoded string? The part I'm struggling to wrap my head around is what exactly am I **generating** in these new lookup methods. A big switch of every single public static property in the `Resource.Drawable` class? Basically this new class should do exactly what the current `DrawableHelper` is doing but it just won't be using Reflection to build the dictionary/switch?

Reply:
Also, where is this being generated? This wouldn't be part of the Android head right? It would have to generate a class that lives somewhere in the Uno assembly? Unless we generate this class in the Android head and add a reference to it to the DrawableHelper.Android.cs partial class, in the same sort of way we do:
https://github.com/unoplatform/uno/blob/0c9de5d418ab197928bda5a69c61670298a18652/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs#L475-L477

We could do something like:
```
writer.AppendLineInvariant($"#if __ANDROID__");
writer.AppendLineInvariant($"global::Uno.Helpers.DrawableHelper.DrawableLookup = new global::{_defaultNamespace}.DrawableLookup());";
writer.AppendLineInvariant($"#endif");
```

Where `.DrawableHelper.DrawableLookup` would be a `IDrawableLookup`, and that interface would live inside Uno and allow us to call the methods without having an implementation inside Uno itself.

Does any of this make sense?